Author: Robin Dong <sanbai@taobao.com>
From: Robin Dong <sanbai@taobao.com>
Date: Thu Nov 29 15:17:51 2012 +0800
Subject: [PATCH 25/26] netoops: dump messages without lock when panic
Patch-mainline:  In house
References: 

    When system is panic, try to dump kernel messages even with out gain the spin_lock.

Signed-off-by: Robin Dong <sanbai@taobao.com>

---
Index: linux-2.6.32-279.9.1.el5/kernel/printk.c
===================================================================
--- linux-2.6.32-279.9.1.el5.orig/kernel/printk.c
+++ linux-2.6.32-279.9.1.el5/kernel/printk.c
@@ -1568,11 +1568,10 @@ void kmsg_dump(enum kmsg_dump_reason rea
 		l2 = chars;
 	}
 
-	if (!spin_trylock_irqsave(&dump_list_lock, flags)) {
-		printk(KERN_ERR "dump_kmsg: dump list lock is held during %s, skipping dump\n",
+	if (!spin_trylock_irqsave(&dump_list_lock, flags))
+		printk(KERN_ERR "dump_kmsg: dump list lock is held during %s, try to dump directly\n",
 				kmsg_to_str(reason));
-		return;
-	}
+
 	list_for_each_entry(dumper, &dump_list, list)
 		dumper->dump(dumper, reason, pt_regs, s1, l1, s2, l2);
 	spin_unlock_irqrestore(&dump_list_lock, flags);
